File: Loenn/entities/SolidcolorEntity.lua
local utils = require("utils")
local drawable = require("structs.drawable_rectangle")

local SolidcolorEntity = {}

SolidcolorEntity.name = "AletrisSandbox/SolidcolorEntity"
SolidcolorEntity.minimumSize = {8,8}
SolidcolorEntity.Depth = -8500
SolidcolorEntity.fieldInformation = {
    width = { fieldType = "integer", default = 16 },
    height = { fieldType = "integer", default = 16 },
    Color = { fieldType = "color",allowXNAColors = true, default = "#FFFFFF" },
    LineColor = { fieldType = "color",allowXNAColors = true, default = "#000000" },
    Interaction = { options = {"None", "Deadly", "Solid"} , editable = false, default = "Solid" },
    DrawType = { options = {"Line","Fill","Bordered"} , editable = false, default = "Bordered" }
}

SolidcolorEntity.fieldOrder = {"x","y","width","height","Color","LineColor","DrawType","Interaction","Depth"}

SolidcolorEntity.placements = {
    name = "Solid Color Entity",
    placementType = "rectangle",
    data = {
        width = 16,
        height = 16,
        Color = "#FFFFFF",
        LineColor = "#000000",
        Interaction = "Solid",
        DrawType = "Bordered"
    }
 }

function SolidcolorEntity.sprite(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    local width, height = entity.width or 8, entity.height or 8
    local color = entity.Color --custom
    local bordercolor = entity.LineColor
    local rect = drawable.fromRectangle(string.lower(entity.DrawType), x, y, width, height, color, bordercolor) --"line","fill","bordered"
    return rect
end

return SolidcolorEntity

File: Source/Entities/SolidcolorEntity.cs
using Celeste.Mod.Entities;
using Microsoft.Xna.Framework;
using Monocle;

namespace Celeste.Mod.AletrisSandbox.Entities;

[CustomEntity("AletrisSandbox/SolidcolorEntity")]
public class SolidcolorEntity : Entity
{
    public Color color;
    public Color BorderColor;

    public enum InteractionType
    {
        None,
        Deadly,
        Solid
    }

    public readonly InteractionType interactions;

    public enum DrawType
    {
        Line,
        Fill,
        Bordered
    }

    public readonly DrawType drawType;
    public new int depth;
    readonly Solid solid;
    readonly bool kill;
    public PlayerCollider pc;

    public SolidcolorEntity(EntityData data, Vector2 offset) : base(data.Position + offset)
    {
        kill = false;
        depth = data.Int("Depth", 8500);
        Depth = depth;
        color = data.HexColor("Color", Calc.HexToColor("#FFFFFF"));
        BorderColor = data.HexColor("LineColor", Calc.HexToColor("#000000"));
        drawType = data.Enum("DrawType", DrawType.Line);
        Collider = new Hitbox(data.Width, data.Height);
        interactions = data.Enum("Interaction", InteractionType.None);
        Add(pc = new(OnCollide));

        switch (interactions)
        {
            case InteractionType.Deadly:
                kill = true;

                break;
            case InteractionType.Solid:
                solid = new(Position, Width, Height, safe: false);

                break;
            case InteractionType.None:
                break;
        }
    }

    void OnCollide(Player player)
    {
        if (OnCollide != null && kill)
            player.Die((player.Center - Center).SafeNormalize());
    }

    public override void Added(Scene scene)
    {
        base.Added(scene);
        if (solid != null)
            scene.Add(solid);
    }

    public override void Render()
    {
        if (drawType == DrawType.Line)
            Draw.HollowRect(Collider, color);
        else if (drawType == DrawType.Fill)
            Draw.Rect(Collider, color);
        else if (drawType == DrawType.Bordered)
        {
            Draw.HollowRect(Collider, color);
            Draw.Rect(Collider, color);
        }
        base.Render();
    }
}